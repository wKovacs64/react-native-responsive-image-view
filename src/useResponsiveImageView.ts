import * as React from 'react';
import {
  Image,
  type ViewProps,
  type ImageProps,
  type ImageURISource,
  type ImageRequireSource,
} from 'react-native';

type ImagePropsWithSourceIgnored = { source?: ImageProps['source'] } & Omit<ImageProps, 'source'>;

export interface ResponsiveImageViewBag {
  loading: boolean;
  error: string;
  retry: () => void;
  getViewProps: (props?: ViewProps) => ViewProps;
  getImageProps: (props?: ImagePropsWithSourceIgnored) => ImageProps;
}

type RenderFunctionReturnType = React.ReactElement<ResponsiveImageViewProps> | null;

export interface ResponsiveImageViewProps {
  aspectRatio?: number;
  component?: React.ComponentType | React.FunctionComponent;
  render?: (bag: ResponsiveImageViewBag) => RenderFunctionReturnType;
  children?: ((bag: ResponsiveImageViewBag) => RenderFunctionReturnType) | React.ReactNode;
  onError?: (errMessage: string) => void;
  onLoad?: () => void;
  source: ImageURISource | ImageRequireSource;
}

// A cancelable version of Image.getSize, adapted from
// https://github.com/kodefox/react-native-flex-image
function getImageSize(
  uri: string,
  onImageSizeSuccess: (width: number, height: number) => void,
  onImageSizeFailure: (err: any) => void,
) {
  let totallyCanceled = false;

  Image.getSize(
    uri,
    (width, height) => {
      if (!totallyCanceled) {
        onImageSizeSuccess(width, height);
      }
    },
    (err) => {
      if (!totallyCanceled) {
        onImageSizeFailure(err);
      }
    },
  );

  return {
    cancel: () => {
      totallyCanceled = true;
    },
  };
}

interface State {
  loading: boolean;
  error: string;
  retryCount: number;
  aspectRatio: number | undefined;
}

type Action =
  | { type: 'SUCCESS'; payload: number }
  | { type: 'FAILURE'; payload: string }
  | { type: 'RETRY' };

const initialState: State = { loading: true, error: '', retryCount: 0, aspectRatio: undefined };

function reducer(state: State, action: Action) {
  switch (action.type) {
    case 'SUCCESS':
      return {
        ...initialState,
        loading: false,
        retryCount: state.retryCount,
        aspectRatio: action.payload,
      };
    case 'FAILURE':
      return {
        ...initialState,
        loading: false,
        error: action.payload,
        retryCount: state.retryCount,
      };
    case 'RETRY':
      return { ...initialState, retryCount: state.retryCount + 1 };
    /* istanbul ignore next: this will never happen */
    default:
      throw new Error('Unexpected action type');
  }
}

function defaultOnLoad() {}
function defaultOnError(_: string) {}

export function useResponsiveImageView({
  aspectRatio: controlledAspectRatio,
  source: initialSource,
  onLoad = defaultOnLoad,
  onError = defaultOnError,
}: ResponsiveImageViewProps): ResponsiveImageViewBag {
  if (!initialSource) {
    throw new Error('"source" is required');
  }

  const [state, dispatch] = React.useReducer(reducer, initialState);

  function retry() {
    dispatch({ type: 'RETRY' });
  }

  function isAspectRatioControlled() {
    return controlledAspectRatio !== undefined;
  }

  function getAspectRatio() {
    return isAspectRatioControlled() ? controlledAspectRatio : state.aspectRatio;
  }

  function getImageProps({
    source,
    style = {},
    ...props
  }: Parameters<ResponsiveImageViewBag['getImageProps']>[0] = {}) {
    const imageProps: ImageProps = {
      source: initialSource,
      style: [style, { height: '100%', width: '100%' }],
      ...props,
    };
    return imageProps;
  }

  function getViewProps({
    style = {},
    ...props
  }: Parameters<ResponsiveImageViewBag['getViewProps']>[0] = {}) {
    return { style: [style, { aspectRatio: getAspectRatio() }], ...props };
  }

  React.useEffect(() => {
    let pendingGetImageSize = { cancel: /* istanbul ignore next: just a stub  */ () => {} };

    function handleImageSizeSuccess(width: number, height: number) {
      onLoad();
      dispatch({ type: 'SUCCESS', payload: width / height });
    }

    function handleImageSizeFailure(err: any) {
      const errMessage =
        err instanceof Error ? err.message : /* istanbul ignore next */ String(err);
      onError(errMessage);
      dispatch({ type: 'FAILURE', payload: errMessage });
    }

    if (typeof initialSource === 'object' && initialSource.uri) {
      // Retrieve image dimensions from URI
      pendingGetImageSize = getImageSize(
        initialSource.uri,
        handleImageSizeSuccess,
        handleImageSizeFailure,
      );
    } else {
      // Retrieve image dimensions from imported resource
      const imageSource = Image.resolveAssetSource(initialSource);
      // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
      if (imageSource) {
        handleImageSizeSuccess(imageSource.width, imageSource.height);
      } else {
        handleImageSizeFailure(new Error('Failed to retrieve image dimensions.'));
      }
    }

    return () => {
      pendingGetImageSize.cancel();
    };
    // Using JSON.stringify here because the `source` parameter can be a nested
    // object. The alternative is requiring the user to memoize it, by why make
    // them do that when we don't have to? (Note: they already have to memoize
    // onLoad and onError, but those are much less likely to be used.)
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [JSON.stringify(initialSource), onLoad, onError, state.retryCount]);

  return { loading: state.loading, error: state.error, retry, getViewProps, getImageProps };
}
