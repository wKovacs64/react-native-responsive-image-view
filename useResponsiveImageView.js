import React from 'react';
import { Image } from 'react-native';

// A cancelable version of Image.getSize, adapted from
// https://github.com/kodefox/react-native-flex-image
function getImageSize(uri, onImageSizeSuccess, onImageSizeFailure) {
  let totallyCanceled = false;

  Image.getSize(
    uri,
    (width, height) => {
      if (!totallyCanceled) {
        onImageSizeSuccess(width, height);
      }
    },
    (err) => {
      if (!totallyCanceled) {
        onImageSizeFailure(err);
      }
    },
  );

  return {
    cancel: () => {
      totallyCanceled = true;
    },
  };
}

const initialState = {
  loading: true,
  error: '',
  aspectRatio: undefined,
};

function reducer(state, action) {
  switch (action.type) {
    case 'SUCCESS':
      return {
        ...initialState,
        loading: false,
        aspectRatio: action.payload,
      };
    case 'FAILURE':
      return {
        ...initialState,
        loading: false,
        error: action.payload,
      };
    /* istanbul ignore next: this will never happen */
    default:
      return state;
  }
}

const noop = () => {};

function useResponsiveImageView({
  aspectRatio: controlledAspectRatio,
  source: initialSource,
  onLoad = noop,
  onError = noop,
}) {
  if (!initialSource) {
    throw new Error('"source" is required');
  }

  const [state, dispatch] = React.useReducer(reducer, initialState);

  function isAspectRatioControlled() {
    return controlledAspectRatio !== undefined;
  }

  function getAspectRatio() {
    return isAspectRatioControlled()
      ? controlledAspectRatio
      : state.aspectRatio;
  }

  function getImageProps({ source, style = {}, ...props } = {}) {
    return {
      source: initialSource,
      style: [style, { height: '100%', width: '100%' }],
      ...props,
    };
  }

  function getViewProps({ style = {}, ...props } = {}) {
    return {
      style: [style, { aspectRatio: getAspectRatio() }],
      ...props,
    };
  }

  React.useEffect(() => {
    let pendingGetImageSize = {
      cancel: /* istanbul ignore next: just a stub  */ () => {},
    };

    function handleImageSizeSuccess(width, height) {
      onLoad();
      dispatch({ type: 'SUCCESS', payload: width / height });
    }

    function handleImageSizeFailure(err) {
      onError(err.message);
      dispatch({ type: 'FAILURE', payload: err.message });
    }

    if (initialSource.uri) {
      // Retrieve image dimensions from URI
      pendingGetImageSize = getImageSize(
        initialSource.uri,
        handleImageSizeSuccess,
        handleImageSizeFailure,
      );
    } else {
      // Retrieve image dimensions from imported resource
      const imageSource = Image.resolveAssetSource(initialSource);
      if (imageSource && imageSource.width && imageSource.height) {
        const { width, height } = imageSource;
        handleImageSizeSuccess(width, height);
      } else {
        handleImageSizeFailure(
          new Error('Failed to retrieve image dimensions.'),
        );
      }
    }

    return () => {
      pendingGetImageSize.cancel();
    };
    // Using JSON.stringify here because the `source` parameter can be a nested
    // object. The alternative is requiring the user to memoize the parameters,
    // but that would add usage overhead and potential confusion (at least in
    // the early days of hooks).
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [JSON.stringify(initialSource), onLoad, onError]);

  return {
    loading: state.loading,
    error: state.error,
    getViewProps,
    getImageProps,
  };
}

export default useResponsiveImageView;
