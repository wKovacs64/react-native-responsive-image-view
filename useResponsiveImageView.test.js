import React from 'react';
import { render } from '@testing-library/react-native';
import {
  mockUriGood,
  mockUriBad,
  mockUriSlowGood,
  mockUriSlowBad,
  mockResourceGood,
  mockResourceBad,
  computedAspectRatio,
  controlledAspectRatio,
  consumerViewProps,
  consumerImageProps,
} from './test/fixtures';
import useResponsiveImageView from './useResponsiveImageView';

const Comp = ({ children, ...props }) =>
  children(useResponsiveImageView(props));

const renderHook = ({ children = jest.fn(() => null), ...props }) => ({
  children,
  ...render(<Comp {...props}>{children}</Comp>),
});

describe('parameter validation', () => {
  it('requires a source', () => {
    // Prevent React from logging errors
    jest.spyOn(console, 'error').mockImplementation(() => {});
    expect(() => renderHook({})).toThrow(/source/);
    // eslint-disable-next-line no-console
    console.error.mockRestore();
  });
});

describe('loading', () => {
  it('is true while loading and otherwise false', async () => {
    const { children } = renderHook({ source: { uri: mockUriGood } });
    expect(children).toHaveBeenCalledWith(
      expect.objectContaining({ loading: true }),
    );
    expect(children).toHaveBeenCalledWith(
      expect.objectContaining({ loading: false }),
    );
  });
});

describe('error', () => {
  it('contains an error message on failure', async () => {
    const { children } = renderHook({ source: { uri: mockUriBad } });
    expect(children).toHaveBeenCalledWith(
      expect.objectContaining({
        error: expect.any(String),
      }),
    );
    expect(children.mock.calls[1][0].error).not.toHaveLength(0);
  });

  it('is empty on success', async () => {
    const { children } = renderHook({ source: { uri: mockUriGood } });
    expect(children).toHaveBeenCalledWith(
      expect.objectContaining({
        error: '',
      }),
    );
  });
});

describe('getViewProps', () => {
  it('includes controlled aspectRatio', async () => {
    const { children } = renderHook({
      aspectRatio: controlledAspectRatio,
      source: { uri: mockUriGood },
    });
    const { getViewProps } = children.mock.calls[1][0];
    expect(getViewProps().style).toContainEqual({
      aspectRatio: controlledAspectRatio,
    });
  });

  it('includes calculated aspectRatio if not provided', async () => {
    const { children } = renderHook({ source: { uri: mockUriGood } });
    const { getViewProps } = children.mock.calls[1][0];
    expect(getViewProps().style).toContainEqual({
      aspectRatio: computedAspectRatio,
    });
  });

  it('composes consumer props with own props', async () => {
    expect.assertions(consumerViewProps.length);
    const { children } = renderHook({ source: { uri: mockUriGood } });
    const { getViewProps } = children.mock.calls[1][0];
    const mergedProps = getViewProps(consumerViewProps);
    Object.keys(consumerViewProps).forEach(consumerProp => {
      expect(mergedProps[consumerProp]).toMatchSnapshot(consumerProp);
    });
  });
});

describe('getImageProps', () => {
  it('sets height and width to 100%', async () => {
    const { children } = renderHook({ source: { uri: mockUriGood } });
    const { getImageProps } = children.mock.calls[1][0];
    expect(getImageProps().style).toContainEqual({
      height: '100%',
      width: '100%',
    });
  });

  it('composes consumer props with own props', async () => {
    expect.assertions(consumerImageProps.length);
    const { children } = renderHook({ source: { uri: mockUriGood } });
    const { getImageProps } = children.mock.calls[1][0];
    const mergedProps = getImageProps(consumerImageProps);
    Object.keys(consumerImageProps).forEach(consumerProp => {
      expect(mergedProps[consumerProp]).toMatchSnapshot(consumerProp);
    });
  });
});

describe('completion callbacks', () => {
  const onLoad = jest.fn();
  const onError = jest.fn();

  it("doesn't throw on success if no onLoad was provided", () => {
    expect(() => renderHook({ source: { uri: mockUriGood } })).not.toThrow();
  });

  it("doesn't throw on failure if no onError was provided", () => {
    expect(() => renderHook({ source: { uri: mockUriBad } })).not.toThrow();
  });

  it('calls provided onLoad on URI success', async () => {
    expect(onLoad).toHaveBeenCalledTimes(0);
    expect(onError).toHaveBeenCalledTimes(0);
    renderHook({ source: { uri: mockUriGood }, onLoad, onError });
    expect(onError).toHaveBeenCalledTimes(0);
    expect(onLoad).toHaveBeenCalledTimes(1);
  });

  it('calls provided onError on URI failure', async () => {
    expect(onLoad).toHaveBeenCalledTimes(0);
    expect(onError).toHaveBeenCalledTimes(0);
    renderHook({ source: { uri: mockUriBad }, onLoad, onError });
    expect(onLoad).toHaveBeenCalledTimes(0);
    expect(onError).toHaveBeenCalledTimes(1);
    expect(onError).toHaveBeenCalledWith(expect.any(String));
  });

  it('calls provided onLoad on imported resource success', async () => {
    expect(onLoad).toHaveBeenCalledTimes(0);
    expect(onError).toHaveBeenCalledTimes(0);
    renderHook({ source: mockResourceGood, onLoad, onError });
    expect(onError).toHaveBeenCalledTimes(0);
    expect(onLoad).toHaveBeenCalledTimes(1);
  });

  it('calls provided onError on imported resource failure', async () => {
    expect(onLoad).toHaveBeenCalledTimes(0);
    expect(onError).toHaveBeenCalledTimes(0);
    renderHook({ source: mockResourceBad, onLoad, onError });
    expect(onLoad).toHaveBeenCalledTimes(0);
    expect(onError).toHaveBeenCalledTimes(1);
    expect(onError).toHaveBeenCalledWith(expect.any(String));
  });

  it('does not call onLoad on success after unmounting', async () => {
    expect(onLoad).toHaveBeenCalledTimes(0);
    expect(onError).toHaveBeenCalledTimes(0);
    const { unmount } = renderHook({
      source: { uri: mockUriSlowGood },
      onLoad,
      onError,
    });
    unmount();
    expect(onError).toHaveBeenCalledTimes(0);
    expect(onLoad).toHaveBeenCalledTimes(0);
  });

  it('does not call onError on failure after unmounting', async () => {
    expect(onLoad).toHaveBeenCalledTimes(0);
    expect(onError).toHaveBeenCalledTimes(0);
    const { unmount } = renderHook({
      source: { uri: mockUriSlowBad },
      onLoad,
      onError,
    });
    unmount();
    expect(onError).toHaveBeenCalledTimes(0);
    expect(onLoad).toHaveBeenCalledTimes(0);
  });
});
