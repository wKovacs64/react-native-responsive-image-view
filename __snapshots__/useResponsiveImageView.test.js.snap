// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getImageProps composes consumer props with own props: onLayout 1`] = `[MockFunction Image#onLayout]`;

exports[`getImageProps composes consumer props with own props: style 1`] = `
Array [
  Object {
    "height": "50%",
    "width": "50%",
  },
  Object {
    "height": "100%",
    "width": "100%",
  },
]
`;

exports[`getViewProps composes consumer props with own props: hitSlop 1`] = `
Object {
  "bottom": 10,
  "left": 0,
  "right": 0,
  "top": 10,
}
`;

exports[`getViewProps composes consumer props with own props: style 1`] = `
Array [
  Object {
    "padding": 20,
  },
  Object {
    "aspectRatio": 1.3333333333333333,
  },
]
`;
